src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -0,0 +1,253 @@
+                                cfs.metadata().name,
+                                cfs.metadata().partitionKeyType,
+                                cfs.metadata().comparator,
+                                expression.column(),
+                                IndexTarget.Type.VALUES,
+                                null);
+    }
+
+    public StorageAttachedIndex getBestIndexFor(RowFilter.Expression expression)
+    {
+        return cfs.indexManager.getBestIndexFor(expression, StorageAttachedIndex.class).orElse(null);
+    }
+
+    public UnfilteredRowIterator getPartition(PrimaryKey key, ReadExecutionController executionController)
+    {
+        if (key == null)
+            throw new IllegalArgumentException("non-null key required");
+
+        try
+        {
+            SinglePartitionReadCommand partition = SinglePartitionReadCommand.create(cfs.metadata(),
+                                                                                     command.nowInSec(),
+                                                                                     command.columnFilter(),
+                                                                                     RowFilter.none(),
+                                                                                     DataLimits.NONE,
+                                                                                     key.partitionKey(),
+                                                                                     makeFilter(key));
+
+            return partition.queryMemtableAndDisk(cfs, executionController);
+        }
+        finally
+        {
+            queryContext.checkpoint();
+        }
+    }
+
+    /**
+     * Build a {@link RangeIterator.Builder} from the given list of expressions by applying given operation (AND).
+     * Building of such builder involves index search, results of which are persisted in the internal resources list
+     *
+     * @param expressions The expressions to build range iterator from (expressions with not results are ignored).
+     *
+     * @return range iterator builder based on given expressions and operation type.
+     */
+    public RangeIterator.Builder getIndexes(Collection<Expression> expressions)
+    {
+        RangeIterator.Builder builder = RangeIntersectionIterator.selectiveBuilder();
+
+        try
+        {
+            for (Expression e : expressions)
+            {
+                if (e.context.isIndexed())
+                {
+                    @SuppressWarnings("resource") // RangeIterators are closed by releaseIndexes
+                    RangeIterator memtableIterator = e.context.searchMemtable(e, mergeRange);
+
+                    builder.add(memtableIterator);
+                }
+            }
+        }
+        catch (Throwable t)
+        {
+            // all sstable indexes in view have been referenced, need to clean up when exception is thrown
+            FileUtils.closeQuietly(builder.ranges());
```

Comment:
nit: This isn't even in the patch, but I think `void closeQuietly(Closeable c)` is actually a duplicate of `void closeQuietly(AutoCloseable c)` :D

Reply:
While I agree I'd strongly prefer not fixing all of these issues in this PR.